fetch/peers/peers.go

```diff
@@ -0,0 +1,148 @@
+package peers
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+type data struct {
+	id                peer.ID
+	success, failures int
+	averageLatency    float64
+}
+
+func (p *data) successRate() float64 {
+	if p.success+p.failures == 0 {
+		return 0
+	}
+	return float64(p.success) / float64(p.success+p.failures)
+}
+
+func (p *data) cmp(other *data) int {
+	if p == nil && other != nil {
+		return -1
+	}
+	const rateThreshold = 0.1
+	switch {
+	case p.successRate()-other.successRate() > rateThreshold:
+		return 1
+	case other.successRate()-p.successRate() > rateThreshold:
+		return -1
+	}
```

Comment:
Would it make sense to keep the `successRate` in the `data` struct and calculate it in the `onLatency()` and `onFailure()` methods to avoid re-calculating it repeatedly in `cmp()`?

Reply:
using a bit more memory (80000 bytes per 10000) for ~30% speed improvement is valid tradeoff